File: leetcode/Merge Two Sorted Lists/1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode curr = null, head = null;
        while(l1 != null && l2 != null){
            int min = 0;
            if(l1.val <= l2.val){
                min = l1.val;
                l1 = l1.next;
            }else{
                min = l2.val;
                l2 = l2.next;
            }
            
            ListNode newNode = new ListNode(min);
            if(curr == null){
                curr = newNode;
                head = curr;
            }else{
                curr.next = newNode;
                curr = curr.next;
            }
        }
        
        while(l1 != null){
            int min = l1.val;
            l1 = l1.next;
            
            ListNode newNode = new ListNode(min);
            if(curr == null){
                curr = newNode;
                head = curr;
            }else{
                curr.next = newNode;
                curr = curr.next;
            }
        }
        while(l2 != null){
            int min = l2.val;
            l2 = l2.next;
            
            ListNode newNode = new ListNode(min);
            if(curr == null){
                curr = newNode;
                head = curr;
            }else{
                curr.next = newNode;
                curr = curr.next;
            }
        }
        
        return head;
    }
}

File: leetcode/Merge Two Sorted Lists/1.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        var list1 = list1
        var list2 = list2
        var head: ListNode? = null
        var curr: ListNode? = null

        while (list1 != null && list2 != null) {
            val min = if (list1.`val` < list2.`val`) {
                val result = list1
                list1 = list1.next
                result
            } else {
                val result = list2
                list2 = list2.next
                result
            }

            if (curr != null) {
                curr.next = min
                curr = min
            } else {
                head = min
                curr = min
            }
        }

        var remaining = list1 ?: list2
        while (remaining != null) {
            val min = remaining
            if (curr != null) {
                curr.next = min
                curr = min
            } else {
                head = min
                curr = min
            }
            remaining = remaining.next
        }
        return head
    }
}